tools/real_time_graphing.py

```diff
@@ -0,0 +1,302 @@
+                    i += 1
+
+            else:
+                color_gen = unique_color_generator([metric['color'] for metric in graph['metrics']])
+
+                # Make axis
+                ax = self.fig.add_subplot(nrows, ncols, len(self.fig.get_axes()) + 1)
+
+                ax.axis([0, 100, 0, 10])
+
+                # Configure the new axis
+                ax.set_title(graph['title'])
+                ax.set_xlabel(graph['xlabel'])
+                ax.set_ylabel(graph['ylabel'])
+
+                for metric in graph['metrics']:
+                    color = metric['color'] if metric['color'] else next(color_gen)
+
+                    m_line, = ax.plot([], [], color=color, label=metric['label'])
+
+                    self.tracked_data.append(Metric(output=m_line, func=metric['func'], x_stream=metric['x_stream'],
+                                                    y_stream=metric['y_stream'], z_stream=metric['z_stream']))
+
+                if graph['legend'] == 'yes' or not graph['legend']:
+                    ax.legend()
+
+    def read_data(self, thread_queue):
+        """
+        Reads data from network and puts it in a queue to be processed.
+
+        Parameters
+        ----------
+        thread_queue: Thread queue
+            Thread queue
+        """
+
+        while not self.thread_stop.is_set():
+            try:
+                data = self.get_data()
+                thread_queue.put(data)
+            except Exception as error:
+                logging.warning("RTG: {}".format(error))
+
+            # Adjust sleep times
+            if self.data_count > self.plot_count:
+                self.sleep_time = self.sleep_time + 1e-5
+
+            elif self.data_count == self.plot_count:
+                self.sleep_time = self.sleep_time - 1e-5
+
+            sleep(self.sleep_time)
+
+    def process_data(self, thread_queue):
+        """
+        Processes data put into the queue.
+
+        Parameters
+        ----------
+        thread_queue: Thread queue
+            Thread queue
+        """
+
+        while not self.thread_stop.is_set():
+            while not thread_queue.empty():
+                data = thread_queue.get(False, self.sleep_time)
+
+                if not data:
+                    logging.warning("RTG: No data!")
+                    sleep(.25)
+                    continue
+
+                self.times.append(time() - self.start_time)
+
+                # Checks data frequency to see if poor quality
+                try:
+                    if self.times[-1] > self.times[-2] + RealTimeGraph.data_freq_warning:
+                        logging.warning("RTG: Data quality: Sucks")
+                        pass
+                except IndexError:
+                    pass
+
+                for metric in self.tracked_data:
+                    func = metric.func
+
+                    x = data[metric.x_stream] if metric.x_stream else None
```

Comment:
Doesn't follow DRY principle. Consider using the following code instead:
```suggestion
                    axes = list('xyz')
                    for axis in axes:
                        stream = getattr(metric, '{}_stream'.format(axis))
                        setattr(self, axis, data[stream] if stream else None)
```